### PR TITLE
Allow to filter `user` by `display_name`

### LINKED ISSUE #766: Admin Enhancement: Make it easier to find specific users
The admin view can be improved in two ways:
- More efficient pagination using UUID based cursors
- Add a search bar to search for a user based on their display name (or prefix).  This will reduce search time when there's over 100 users registered.

### PR DESCRIPTION
**Pull request by notmd**
close #766
~~DO NOT MERGE, depend on #851~~ It's ready now.
- Implement a new `DataTable` component. I implement our own `filter` and `paginate` logic cause the built-in doesn't really useful when fetching data from server side and is inflexible.
- Use `react-table` v8. I plan to migrate other components currently using v7 to v8 when I complete #766.
- Allow searching for the user by `display_name`.
- ~~Use UUID based cursor.~~ It now provided by the backend.
- ~~Fix a bug when querying the user in the backward direction.~~

~~Note that this PR doesn't support filtering users by prefix and the `api/v1/frontend_users/by_display_name` endpoint doesn't support pagination yet. I'm going to address this in the next PR.~~
Here is the new UI

image

### CODE REVIEW

File: website/src/pages/api/admin/users.ts [hunk 1]
```diff
@@ -1,11 +1,12 @@
 import { withRole } from "src/lib/auth";
 import { oasstApiClient } from "src/lib/oasst_api_client";
 import prisma from "src/lib/prismadb";
+import { BackendUser } from "src/types/Users";
 
 /**
  * The number of users to fetch in a single request.  Could later be a query parameter.
  */
-const PAGE_SIZE = 20;
+const PAGE_SIZE = 1;
```
Comment (fozziethebeat) [inline website/src/pages/api/admin/users.ts:9]: Is this part intentional? Or should we switch it back to 20?

**Review by fozziethebeat — CHANGES_REQUESTED** [review]

**Comment by fozziethebeat** [comment]
I like the look of the new table!

File: website/src/pages/api/admin/users.ts (hunk 1 above)
Comment (notmd) [reply website/src/pages/api/admin/users.ts]: Sorry, I was debugging the pagination. Fixed with latest commit

File: backend/oasst_backend/user_repository.py [hunk 2]
```diff
@@ -161,10 +161,10 @@ def query_users(
         users = users.order_by(User.display_name)
 
         if gt:
-            users = users.filter(User.display_name > gt)
+            users = users.filter(User.id > gt)
```
Comment (andreaskoepf) [inline backend/oasst_backend/user_repository.py:164]: Not sure what the intention here is, but it doesn't look right.

Comment (notmd) [reply backend/oasst_backend/user_repository.py:164]: I thought we want to cursor based on the user id?

Comment (notmd) [reply backend/oasst_backend/user_repository.py:164]: Is the user id column monotonically increasing?

Comment (fozziethebeat) [reply backend/oasst_backend/user_repository.py:164]: UUID V4 isn't guaranteed to be monotonically increasing.  I think only UUID V6 or V7 have that property.  V4 is random in it's ordering.

**Review by fozziethebeat — APPROVED** [review]

File: website/src/components/DataTable.tsx [hunk 3]
```diff
@@ -0,0 +1,158 @@
...
+            <Thead>
+              {getHeaderGroups().map((headerGroup) => (
+                <Tr key={headerGroup.id}>
+                  {headerGroup.headers.map((header) => (
+                    <Th key={header.id}>
+                      <Box display="flex" alignItems="center">
+                        {header.isPlaceholder ? null : flexRender(header.column.columnDef.header, header.getContext())}
+                        {(header.column.columnDef as DataTableColumnDef<T>).filterable && (
+                          <FilterModal
+                            value={filterValues.find((value) => value.id === header.id)?.value ?? ""}
```
Comment (rjmacarthy) [inline website/src/components/DataTable.tsx:97]: If `filterValues` was an object and not an array could we avoid using `find` and `findIndex`?

File: website/src/components/UserTable.tsx [hunk 4]
```diff
@@ -0,0 +1,133 @@
...
+  useSWR(
+    `/api/admin/users?direction=${pagination.direction}&cursor=${pagination.cursor}&display_name=${display_name}`,
+    get,
+    {
+      onSuccess: (data) => {
+        // When no more users can be found, trigger a toast to indicate why no
+        // changes have taken place.  We have to maintain a non-empty set of
+        // users otherwise we can't paginate using a cursor (since we've lost the
+        // cursor).
+        if (data.length === 0) {
```
Comment (rjmacarthy) [inline website/src/components/UserTable.tsx:76]:
```suggestion
        if (!data.length) {
```

File: website/src/components/DataTable.tsx (hunk 3 above)
Comment (rjmacarthy) [reply website/src/components/DataTable.tsx:97]: Thank you,

I was mainly referring to a suggestion to use object for items and not an array which would enable memorable look up of filters.

```
const filters = {
  linkOperator: 'is',
  items: {
    1: {
      id: 1,
      columnField: 'rating',
      operatorValue: '>',
      value: '4',
    },
    2: {
      id: 2,
      columnField: 'isAdmin',
      operatorValue: 'is',
      value: 'true',
    },
  },
}

const getFilterById = (id) => filters.items[id]
```

File: backend/oasst_backend/user_repository.py [hunk 5]
```diff
@@ -207,6 +207,7 @@ def query_users_ordered_by_display_name(
         limit: Optional[int] = 100,
         desc: bool = False,
     ) -> list[User]:
+
```
Comment (yk) [inline backend/oasst_backend/user_repository.py:210]: this newline triggers my review as code owner. it's a very beautiful newline. I approve ;)

**Review by yk — APPROVED** [review]

**Review by fozziethebeat — APPROVED** [review]
Time to merge and see all the glory

**Pull request merged**